### PR TITLE
Automation: KubeMacPool functests on CNAO lane

### PR DESCRIPTION
**Pull request by ormergi**
**What this PR does / why we need it**:
This PR adds new test lane, KubeMacPool functional tests on a cluster with latest changes on CNAO,
this also adds the infrastructure that enable one create such automation for any other CNAO component e.g: NMstate

Project-Infra prow job for this lane PR https://github.com/kubevirt/project-infra/pull/457
**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: automation/check-patch.setup.sh [hunk 1]
```diff
@@ -25,3 +25,7 @@ cp -rf $(pwd)/. ${TMP_PROJECT_PATH}
 
 echo 'Exporting temporary project path'
 export TMP_PROJECT_PATH
+
+echo 'Exporting KubeMacPool component project path'
+KUBEMACPOOL_TMP_PROJECT_PATH=${TMP_PROJECT_PATH}/_components/go/src/github.com/k8snetworkplumbingwg/kubemacpool/
```
Comment (phoracek) [inline automation/check-patch.setup.sh:30]: This should not be in the shared setup script as it is irrelevant for most of the targets.

File: automation/check-patch.setup.sh [hunk 2]
```diff
@@ -4,9 +4,9 @@
 # cd ${TMP_PROJECT_PATH}
 
 echo 'Setup Go paths'
-export GOROOT=/tmp/cluster-network-addons-operator/go/root
+export GOROOT=~/tmp/cluster-network-addons-operator/go/root
```
Comment (phoracek) [inline automation/check-patch.setup.sh:7]: I don't want tmp to appear in my home directory. Why did you change it?

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash -xe
+
+# This script should be able to execute KubeMacPool
+# functional tests against Kubernetes cluster on any
+# environment with basic dependencies listed in
+# check-patch.packages installed and docker running.
+#
+# yum -y install automation/check-patch.packages
+# automation/check-patch.e2e-workflow-k8s.sh
```
Comment (phoracek) [inline automation/check-patch.e2e-kubemacpool-functests.sh:9]: Wrong path

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    make cluster-down
+    make cluster-up
+    trap teardown EXIT SIGINT SIGTERM SIGSTOP
+    make cluster-operator-push
+    make cluster-operator-install
+
+    cd ${KUBEMACPOOL_TMP_PROJECT_PATH}
+    export KUBEVIRT_PROVIDER=external
+    export CNAO_PROVIDER=external
+    export KUBECONFIG=${TMP_PROJECT_PATH}/_kubevirtci/_ci-configs/${KUBEVIRT_PROVIDER}/.kubeconfig
```
Comment (phoracek) [inline automation/check-patch.e2e-kubemacpool-functests.sh:30]: You can't depend on private generate directories here. Use https://github.com/kubevirt/cluster-network-addons-operator/blob/master/cluster/kubevirtci.sh#L34 instead

File: automation/check-patch.setup.sh (hunk 1 above)
Comment (phoracek) [inline automation/check-patch.setup.sh:30]: Also, you can't depend on _components, you don't create it here. If you want to use kubemacpool sources here, we need to introduce a way to fetch it explicitly

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    make cluster-down
+    make cluster-up
+    trap teardown EXIT SIGINT SIGTERM SIGSTOP
+
+    # Deploy CNAO with latest changes
+    make cluster-operator-push
+    make cluster-operator-install
+
+    # Deploy Kubevirt
+    export KUBEVIRT_VERSION=v0.29.0
```
Comment (qinqon) [inline automation/check-patch.e2e-kubemacpool-functests.sh:32]: Maybe we should take latest released kubevirt since so we streamline everything, or at lastet latest from a specific major.minor so we get bugfixed release.

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    ./hack/deploy-kubevirt.sh
+
+    # Export kubeconfig absolute path
+    export KUBECONFIG=$(_kubevirtci/cluster-up/kubeconfig.sh)
+
+    # Clone KubeMacPool repo
+    yum install -y shyaml
+    KUBEMACPOOL_PROJECT_PATH=${GOPATH}/src/github.com/k8snetworkplumbingwg/kubemacpool
```
Comment (qinqon) [inline automation/check-patch.e2e-kubemacpool-functests.sh:40]: since we use modules we no longer need a GOPATH stored project we can create a random temporal directory that we remove after we finish `mktemp -d` then a teardown we remove it catching signals.

In fact this way we really tests modules.

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    KUBEMACPOOL_URL=$(cat components.yaml | shyaml get-value components.kubemacpool.url)
+    KUBEMACPOOL_COMMIT=$(cat components.yaml | shyaml get-value components.kubemacpool.commit)
+    mkdir -p $KUBEMACPOOL_PROJECT_PATH
+    git clone $KUBEMACPOOL_URL $KUBEMACPOOL_PROJECT_PATH
+    cd $KUBEMACPOOL_PROJECT_PATH
+    git fetch origin
+    git reset --hard
+    git checkout ${KUBEMACPOOL_COMMIT}
```
Comment (qinqon) [inline automation/check-patch.e2e-kubemacpool-functests.sh:48]: this three lines can be replaced with 
```
git reset ${KUBEMACPOOL_COMMIT} --hard
```

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    # Deploy NetworkAddonsConfig CR
+    export KUBEVIRT_PROVIDER=external
```
Comment (qinqon) [inline automation/check-patch.e2e-kubemacpool-functests.sh:51]: Let's create a function for the KMP tier1 run so KUBEVIRT_PROVIDER does not affect the CNAO kubevirtci steps.

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 5 above)
Comment (RamLavi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:32]: If we use deploy-kubevirt.sh then we need to maintain this (for example bumping the kubevirt version) here and also on KMP repo.
I suggest to drop this script and instead run kubevirt deployment via the kubemacpool repo after cloning:
```
KUBEVIRT_PROVIDER=extarnal KUBECONFIG=$(_kubevirtci/cluster-up/kubeconfig.sh) make cluster-up
```

This way we don't need to maintain 2 KUBEVIRT_VERSION references and only need to bump via KMP repo

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  imagePullPolicy: Always
+  kubeMacPool: {}
+  linuxBridge: {}
+  multus: {}
+EOF
+    ./cluster/kubectl.sh apply -f cr.yaml
+    ./cluster/kubectl.sh wait networkaddonsconfig cluster --for condition=Available --timeout=800s
+
+    # Run KubeMacPool functional tests
+    make functest
```
Comment (RamLavi) [inline automation/check-patch.e2e-kubemacpool-functests.sh:67]: please add ` E2E_TEST_ARGS="--ginkgo.skip='mac-pool is full'  make functest"` since we want to avoid tests using full mac-pool ([KMP PR](https://github.com/k8snetworkplumbingwg/kubemacpool/pull/214) ).

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 5 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:32]: Which script to drop?

**Comment by ormergi** [comment]
@phoracek 
Why did you closed this?

**Comment by phoracek** [comment]
Was cleaning up old WIP PRs. This may be indeed still useful.

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 5 above)
Comment (phoracek) [reply automation/check-patch.e2e-kubemacpool-functests.sh:32]: I agree we should keep this on latest to lower the maintenance cost. Also, we should be compatible with the latest KubeVirt, testing old one would not really help us with integration testing.

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    # Deploy CNAO with latest changes
+    make cluster-operator-push
+    make cluster-operator-install
+
+    # Deploy Kubevirt
+    export KUBEVIRT_VERSION=v0.29.0
+    ./hack/deploy-kubevirt.sh
+
+    # Export kubeconfig absolute path
+    export KUBECONFIG=$(_kubevirtci/cluster-up/kubeconfig.sh)
```
Comment (phoracek) [inline automation/check-patch.e2e-kubemacpool-functests.sh:36]: Please introduce kubeconfig script similar to https://github.com/nmstate/kubernetes-nmstate/blob/master/cluster/kubeconfig.sh. We should not depend on internal "private" scripts here.

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+    # Clone KubeMacPool repo
+    yum install -y shyaml
+    KUBEMACPOOL_PROJECT_PATH=${GOPATH}/src/github.com/k8snetworkplumbingwg/kubemacpool
+    KUBEMACPOOL_URL=$(cat components.yaml | shyaml get-value components.kubemacpool.url)
```
Comment (phoracek) [inline automation/check-patch.e2e-kubemacpool-functests.sh:41]: Ram added helper scripts for this. Please use them.

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    KUBEMACPOOL_COMMIT=$(cat components.yaml | shyaml get-value components.kubemacpool.commit)
+    mkdir -p $KUBEMACPOOL_PROJECT_PATH
+    git clone $KUBEMACPOOL_URL $KUBEMACPOOL_PROJECT_PATH
```
Comment (phoracek) [inline automation/check-patch.e2e-kubemacpool-functests.sh:44]: I think that even this is in those helper scripts.

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 5 above)
Comment (RamLavi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:32]: I think we should use KMP's new function [getLatestPatchVersion](https://github.com/k8snetworkplumbingwg/kubemacpool/blob/master/cluster/up.sh#L19)

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 11 above)
Comment (RamLavi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:41]: [utils](https://github.com/kubevirt/cluster-network-addons-operator/blob/master/hack/components/yaml-utils.sh) and [use example](https://github.com/kubevirt/cluster-network-addons-operator/blob/a98e1d44fba380dc9982f0fe60c5b1c8f796ec82/hack/components/bump-ovs-cni.sh#L36)

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 12 above)
Comment (RamLavi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:44]: [here](https://github.com/kubevirt/cluster-network-addons-operator/blob/a98e1d44fba380dc9982f0fe60c5b1c8f796ec82/hack/components/git-utils.sh#L5)

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 7 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:48]: I will use git-utils.sh

**Comment by ormergi** [comment]
@phoracek @qinqon @RamLavi 
I will address all reviews in following commits,
they will also contain separation of common necessary logic 
as preparation for other components e2e-functests jobs.

File: automation/check-patch.e2e.packages [hunk 13]
```diff
@@ -1,3 +1,4 @@
 docker
 git
 make
+shyaml
```
Comment (oshoval) [inline automation/check-patch.e2e.packages:4]: why ?
at the end Ram used containerized version for it, can you do the same please if possible?

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  imagePullPolicy: Always
+  kubeMacPool: {}
+  linuxBridge: {}
+  multus: {}
+EOF
+    kubectl apply -f cr.yaml
+    kubectl wait networkaddonsconfig cluster --for condition=Available --timeout=13m
+
+    # Run KubeMacPool functional tests
+     KUBECONFIG=${KUBECONFIG} E2E_TEST_ARGS="--ginkgo.skip='mac-pool is full'" make functest
```
Comment (oshoval) [inline automation/check-patch.e2e-kubemacpool-functests.sh:41]: nit: ident

File: automation/e2e-functests.setup.sh [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+source hack/components/git-utils.sh
+source hack/components/yaml-utils.sh
+
+source automation/check-patch.setup.sh
+
+teardown() {
+    #make cluster-down
+    # rm -rf "$kubevirt_temp_dir"
+    # rm -rf "$component_temp_dir"
```
Comment (oshoval) [inline automation/e2e-functests.setup.sh:32]: remove comments please if possible

File: automation/e2e-functests.setup.sh [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    echo hi
+}
+
+trap teardown EXIT SIGINT SIGTERM SIGSTOP
```
Comment (oshoval) [inline automation/e2e-functests.setup.sh:36]: we should add SIGQUIT

File: automation/e2e-functests.setup.sh [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+# Export .kubeconfig full path, so it will be possible
+# to use 'kubectl' directly from the component directory path
+export KUBECONFIG=$(./cluster/kubeconfig.sh)
+
+#Deploy CNAO latest changes
+make cluster-operator-push
+make cluster-operator-install
+
+# Deploy KubeVirt latest stable release
+export KUBEVIRT_VERSION=$(curl -s https://api.github.com/repos/kubevirt/kubevirt/releases | grep tag_name | grep -v -- - | sort -V | tail -1 | awk -F':' '{print $2}' | sed 's/,//' | xargs)
```
Comment (oshoval) [inline automation/e2e-functests.setup.sh:56]: last time i checked the site, the suggested method was
`export KUBEVIRT_VERSION=$(curl -s https://api.github.com/repos/kubevirt/kubevirt/releases|grep tag_name|sort -V | tail -1 | awk -F':' '{print $2}' | sed 's/,//' | xargs)`

is the one you use is the official suggested one ?

Comment (ormergi) [reply automation/e2e-functests.setup.sh:56]: Yes i took it from here https://kubevirt.io/quickstart_minikube/
the difference is `grep -v -- - `, eventually it will skip `rc` versions

Comment (oshoval) [reply automation/e2e-functests.setup.sh:56]: i think the official site has it as well ? (which is preferred to look on imo)
one time it changed the API and changed the suggested method

File: automation/e2e-functests.setup.sh (hunk 16 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:36]: Done

File: automation/e2e-functests.setup.sh (hunk 15 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:32]: Done

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 14 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:41]: What do you mean?

File: hack/deploy-kubevirt.sh [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+count=0
+until ./cluster/kubectl.sh -n kubevirt get kv kubevirt; do
+    ((count++)) && ((count == 30)) && echo "KubeVirt CR not found" && exit 1
+    echo "waiting for KubeVirt CR"
+    sleep 1
+done
+
+./cluster/kubectl.sh wait -n kubevirt kv kubevirt --for condition=Available --timeout 180s || (echo "KubeVirt not ready in time" && exit 1)
+
+echo "Done"
```
Comment (oshoval) [inline hack/deploy-kubevirt.sh:26]: suggestion echo the script name + Done
maybe `basename -- "$0"` can be used?

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 14 above)
Comment (oshoval) [reply automation/check-patch.e2e-kubemacpool-functests.sh:41]: line 41 has one space more than it needs, isnt it ?

Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:41]: Gotcha, I will fix it in the following commits

File: automation/check-patch.e2e.packages (hunk 13 above)
Comment (ormergi) [reply automation/check-patch.e2e.packages:4]: Can you refere me to that code/script?
FYI, all scripts under automation are scripts that run on prow

File: hack/deploy-kubevirt.sh (hunk 18 above)
Comment (ormergi) [reply hack/deploy-kubevirt.sh:26]: I agree, this should be more informative

File: automation/check-patch.e2e.packages (hunk 13 above)
Comment (oshoval) [reply automation/check-patch.e2e.packages:4]: https://github.com/kubevirt/cluster-network-addons-operator/blob/42c44712947d5841ec4623279e84ad31cf27da57/hack/components/yaml-utils.sh#L5

doesnt matter where they run, this one will be generic

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 14 above)
Comment (RamLavi) [inline automation/check-patch.e2e-kubemacpool-functests.sh:41]: can you make sure that `ginkgo.skip` works as expected so that these tests don't run on production?

File: automation/e2e-functests.setup.sh [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+./hack/deploy-kubevirt.sh
```
Comment (phoracek) [inline automation/e2e-functests.setup.sh:57]: This should not be in the shared functest.setup. The only component where we make use of KubeVirt is KubeMacPool. This it should move to KMP script.

File: automation/e2e-functests.setup.sh [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+teardown() {
+    #make cluster-down
+    # rm -rf "$kubevirt_temp_dir"
+    # rm -rf "$component_temp_dir"
+    echo hi
+}
+
+trap teardown EXIT SIGINT SIGTERM SIGSTOP
+
+export COMPONENT=${COMPONENT:-kubemacpool}
```
Comment (RamLavi) [inline automation/e2e-functests.setup.sh:38]: It seems you moved some common stuff to `e2e-functests.setup.sh`, similar to `check-patch.setup.sh`, right?. 
If so, then it's common so please remove traces of kubemacpool from it.

**Comment by phoracek** [comment]
/hold

Let's add an optional job to kubevirt-infra for this, so we can test it on the PR.

**Comment by RamLavi** [comment]
> Let's add an optional job to kubevirt-infra for this, so we can test it on the PR.

I think @ormergi already opened a [PR](https://github.com/kubevirt/project-infra/pull/457) on project infra

**Comment by phoracek** [comment]
Great, thanks

**Comment by phoracek** [comment]
@ormergi 
```
automation/e2e-functests.setup.sh: line 49: ./cluster/kubeconfig.sh: No such file or directory
++ export KUBECONFIG=
```

**Comment by ormergi** [comment]
> @ormergi
> 
> ```
> automation/e2e-functests.setup.sh: line 49: ./cluster/kubeconfig.sh: No such file or directory
> ++ export KUBECONFIG=
> ```
Thanks for the heads up I will fix this.

Calling `check-patch.setup.sh` from `e2e-functests-setup.sh` wasn't a good idea,
using `cd` inside script called from another script (at `e2e-functests-setup.sh`) is confusing.
The following commits untangle this awkward structure so `check-patch.e2e-kubemacpool-functests.sh` will call each of the mentioned script sequentially and move to the right directories.
I will also change the name of the script `e2e-functests-setup.sh` to `components-functests.sh` 
it describes better what it does.
And of-course resolve more reviews.

**Comment by phoracek** [comment]
>  I will also change the name of the script e2e-functests-setup.sh to components-functests.sh
>  it describes better what it does.

@ormergi setup` suffix signalized that it is a helper script preparing the environment. Please name it `components-functests.setup.sh` or something similar with `setup.sh` at the end to follow other helper scripts.

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 14 above)
Comment (qinqon) [reply automation/check-patch.e2e-kubemacpool-functests.sh:41]: What about running tier1 tests here ? meaning filtering by polarion id.

File: automation/check-patch.e2e.packages (hunk 13 above)
Comment (qinqon) [reply automation/check-patch.e2e.packages:4]: Prow does not use this file, this is for the old stdci CI system.

File: automation/e2e-functests.setup.sh (hunk 16 above)
Comment (qinqon) [inline automation/e2e-functests.setup.sh:36]: EXIT is enough

File: automation/e2e-functests.setup.sh (hunk 17 above)
Comment (qinqon) [inline automation/e2e-functests.setup.sh:56]: Move this to deploy-kubevirt.sh

File: automation/e2e-functests.setup.sh [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+# Deploy KubeVirt latest stable release
+export KUBEVIRT_VERSION=$(curl -s https://api.github.com/repos/kubevirt/kubevirt/releases | grep tag_name | grep -v -- - | sort -V | tail -1 | awk -F':' '{print $2}' | sed 's/,//' | xargs)
+./hack/deploy-kubevirt.sh
+
+# Clone component repository
+component_url=$(yaml-utils::get_component_url ${COMPONENT})
+component_commit=$(yaml-utils::get_component_commit ${COMPONENT})
+component_repo=$(yaml-utils::get_component_repo ${component_url})
+
+component_temp_dir=$(git-utils::create_temp_path ${COMPONENT})
```
Comment (qinqon) [inline automation/e2e-functests.setup.sh:64]: You have to remove this tmp dir on trap

File: hack/components/git-utils.sh [hunk 22]
```diff
@@ -18,6 +18,17 @@ function git-utils::fetch_component() {
     )
 }
 
+function git-utils::fetch_component_master() {
```
Comment (qinqon) [inline hack/components/git-utils.sh:21]: This is not used

File: hack/deploy-kubevirt.sh [hunk 23]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
Comment (qinqon) [inline hack/deploy-kubevirt.sh:1]: Since we are going to clone kubemacpool can we use the deploy-kubevirt.sh from kubemacpool ? in fact I think we can even use automation from kubemacpool instead of rewritting stuff here.

File: automation/e2e-functests.setup.sh [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+    #make cluster-down
+    # rm -rf "$kubevirt_temp_dir"
+    # rm -rf "$component_temp_dir"
+    echo hi
+}
+
+trap teardown EXIT SIGINT SIGTERM SIGSTOP
+
+export COMPONENT=${COMPONENT:-kubemacpool}
+
```
Comment (qinqon) [inline automation/e2e-functests.setup.sh:39]: Since we are going to clone kubemacpool I think we can reuse `kubemacpool/automation` stuff.

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 14 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh:41]: @RamLavi  I passed `-ginkgo.skip=mac-pool is full` it skipped those test

@qinqon I think we should run all tier1 test shouldnt we?

File: automation/check-patch.e2e.packages (hunk 13 above)
Comment (ormergi) [reply automation/check-patch.e2e.packages:4]: @qinqon @oshoval 
Got it, I will remove this change

File: automation/e2e-functests.setup.sh (hunk 16 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:36]: Ok

File: automation/e2e-functests.setup.sh (hunk 20 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:38]: Will do

File: automation/e2e-functests.setup.sh (hunk 24 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:39]: I use CNAO cluster-up since it deploys extra packages on the cluster nodes,
we make use `make functest` of kubemacpool though

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 14 above)
Comment (phoracek) [reply automation/check-patch.e2e-kubemacpool-functests.sh:41]: > @RamLavi I passed -ginkgo.skip=mac-pool is full it skipped those test

@ormergi why do we skip them? We can specify (limit) the size of the MAC range in the CR.

File: automation/e2e-functests.setup.sh (hunk 17 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:56]: I add `-u` flag for `deploy-kubevirt.sh` it is mandatory to export KUBEVIRT_VERSION,
and this is the version that will be deployed.
I did not put this in that script because I wanted to allow deploying specific version and in case we want the latest 
the use should figure which version is that and pass it.

How about adding a logic that will check if `KUBEVIRT_VERSION=latest-stable` then we use that `curl..` command?

File: automation/e2e-functests.setup.sh (hunk 19 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:57]: Will do

File: automation/e2e-functests.setup.sh (hunk 21 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:64]: I will change this in following commits

File: automation/e2e-functests.setup.sh (hunk 17 above)
Comment (oshoval) [reply automation/e2e-functests.setup.sh:56]: if i understood you right, you can use the method to get latest as you did, unless
KUBEVIRT_VERSION already has value given by the user

File: hack/deploy-kubevirt.sh (hunk 23 above)
Comment (ormergi) [reply hack/deploy-kubevirt.sh:1]: Currently we dont have it on kubemacpool, should I move it to there?

**Comment by ormergi** [comment]
> > I will also change the name of the script e2e-functests-setup.sh to components-functests.sh
> > it describes better what it does.
> 
> @ormergi setup`suffix signalized that it is a helper script preparing the environment. Please name it`components-functests.setup.sh`or something similar with`setup.sh` at the end to follow other helper scripts.

 No problem I will change it to `components-functests.setup.sh`

Should we run kubemacpool tests from kubemacpool master branch or from the commit that mentioned in components.yaml?

**Comment by phoracek** [comment]
> Should we run kubemacpool tests from kubemacpool master branch or from the commit that mentioned in components.yaml?

Good question. It should be from the commit.

File: automation/e2e-functests.setup.sh (hunk 17 above)
Comment (ormergi) [reply automation/e2e-functests.setup.sh:56]: Right

File: hack/components/git-utils.sh (hunk 22 above)
Comment (ormergi) [reply hack/components/git-utils.sh:21]: I will remove it since we need a specific commit and not the latest master

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 14 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh]: Ram suggested to skip to tests because they might stuck, that comment vanished because I pushed new commit.

When kubemacpool deployed by CNAO, it is managed by CNAO so if we change the macpool range value and restart KMP pods, CNAO will bring KMP pods back with the previous macpool range values.
@RamLavi  is that right?

Comment (phoracek) [reply automation/check-patch.e2e-kubemacpool-functests.sh]: I think we don't control the range by KMP e2e anymore, do we?

**Comment by ormergi** [comment]
We got results! 
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_cluster-network-addons-operator/408/pull-e2e-cluster-network-addons-operator-kubemacpool-functests/1290669747393794048

The tests are working some even fail and we get JUnit report

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 14 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh]: You right, we dont change the range in e2e anymore 
we do restart KMP pods by manipulating KMP deployment replica number.
I will verify "full macpool" tests are passing, if they are I will remove skip flag.

**Comment by ormergi** [comment]
New commit changes:
Set mac address range in NetworkAaddonsConfig and skip flag in kubemacpool tests suite.
Improve script description

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+   rangeEnd: "02:00:00:00:00:0F"
+  linuxBridge: {}
+  multus: {}
+EOF
+
+    kubectl wait networkaddonsconfig cluster --for condition=Available --timeout=13m
+
+    # Run KubeMacPool functional tests
+    cd ${TMP_COMPONENT_PATH}
+    KUBECONFIG=${KUBECONFIG} E2E_TEST_ARGS="--junit-output=$ARTIFACTS/junit.functest.xml" make functest
```
Comment (qinqon) [inline automation/check-patch.e2e-kubemacpool-functests.sh:51]: This is not always true, maybe is good that we align the components.

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+main() {
+    # Setup CNAO and artifacts temp directory
+    source automation/check-patch.setup.sh
+    cd ${TMP_PROJECT_PATH}
+
+    # Spin-up ephemeral cluster with latest CNAO
+    # this script also exports KUBECONFIG, and fetch $COMPONENT repository
+    export COMPONENT="kubemacpool"
```
Comment (phoracek) [inline automation/check-patch.e2e-kubemacpool-functests.sh:24]: We don't need to export this, we can instead just call following line with `COMPONENT='kubemacpool' source automation/components-functests.setup.sh`.

That way we make sure the export does not spoil the rest of the scope or other called scripts.

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+# functional tests against Kubernetes cluster with
+# CNAO built with latest changes, on any
+# environment with basic dependencies listed in
+# check-patch.packages installed and docker running.
+#
+# yum -y install automation/check-patch.packages
+# automation/check-patch.e2e-kubemacpool-functests.sh
+
+teardown() {
+    echo "make cluster-down"
```
Comment (phoracek) [inline automation/check-patch.e2e-kubemacpool-functests.sh:13]: echo?

File: automation/check-patch.e2e-kubemacpool-functests.sh [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+    cd ${TMP_PROJECT_PATH}
+
+    # Spin-up ephemeral cluster with latest CNAO
+    # this script also exports KUBECONFIG, and fetch $COMPONENT repository
+    export COMPONENT="kubemacpool"
+    source automation/components-functests.setup.sh
+
+    trap teardown EXIT
+
+    # Deploy KubeVirt latest stable release
```
Comment (phoracek) [inline automation/check-patch.e2e-kubemacpool-functests.sh:29]: Make it an echo, so it makes for a nice log

File: automation/components-functests.setup.sh [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+# - KUBECONFIG
+#   kubeconfig full path so you can use`kubectl` binary or
+#   CNAO's `./cluster/kubectl.sh` script directly.
+#
+# - TMP_COMPONENT_PATH
+#   The desired component temp directory
+#
+# Example:
+# export COMPONENT="KubeMacPool"
+# source automation/e2e-functest.setup.sh
```
Comment (phoracek) [inline automation/components-functests.setup.sh:17]: Again, COMPONENT can be set next to `source` to limit the scope

File: automation/components-functests.setup.sh [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+
+# Spin up Kubernetes cluster
+export KUBEVIRT_PROVIDER='k8s-1.17'
+make cluster-down cluster-up
+
+# Export .kubeconfig full path, so it will be possible
+# to use 'kubectl' directly from the component directory path
+export KUBECONFIG=$(kubevirtci::kubeconfig)
+
+#Deploy CNAO latest changes
```
Comment (phoracek) [inline automation/components-functests.setup.sh:40]: `# Deploy`

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 28 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh]: Done

File: automation/components-functests.setup.sh (hunk 29 above)
Comment (ormergi) [reply automation/components-functests.setup.sh]: Done

File: automation/components-functests.setup.sh (hunk 30 above)
Comment (ormergi) [reply automation/components-functests.setup.sh]: Done

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 26 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh]: Done

File: hack/deploy-kubevirt.sh [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+
+# Ensure the KubeVirt CR is created
+count=0
+until ./cluster/kubectl.sh -n kubevirt get kv kubevirt; do
+    ((count++)) && ((count == 30)) && echo "KubeVirt CR not found" && exit 1
+    echo "waiting for KubeVirt CR"
+    sleep 1
+done
+
+./cluster/kubectl.sh wait -n kubevirt kv kubevirt --for condition=Available --timeout 180s || (echo "KubeVirt not ready in time" && exit 1)
```
Comment (phoracek) [inline hack/deploy-kubevirt.sh:29]: 3 minutes feels a little tight for the deployment of the whole KubeVirt. Does this work ok on KMP tests?

File: automation/check-patch.e2e-kubemacpool-functests.sh (hunk 27 above)
Comment (ormergi) [reply automation/check-patch.e2e-kubemacpool-functests.sh]: Forgot to remove it while debugging sorry :x

File: hack/deploy-kubevirt.sh (hunk 31 above)
Comment (ormergi) [reply hack/deploy-kubevirt.sh:29]: Yep seems fine until now

**Comment by ormergi** [comment]
New commit changes:
Now CNAO deploys all components, this part also moved to `component-functests.setup.sh`
Resolved nits

**Comment by ormergi** [comment]
/unhold
PR is tidy now and we have kubevirt-infra job for this new lane.

**Comment by phoracek** [comment]
/approve

@ormergi would you please open a PR on infra to make the KMP lane run on all PRs and visible. Just keep it optional, so it does not vote.

**Comment by ormergi** [comment]
> Just keep it optional, so it does not vote.

https://github.com/kubevirt/project-infra/pull/571
